model/project_vars.go

```diff
@@ -441,46 +466,123 @@ func (projectVars *ProjectVars) MergeWithRepoVars(repoVars *ProjectVars) {
-
-// getParamNameForVar returns the corresponding parameter name for a project
-// variable. If the project variable does not yet have a parameter name, it
-// generates one.
-func getParamNameForVar(varsToParams []ParameterMapping, varName string) (string, error) {
-	for _, varToParam := range varsToParams {
-		if varToParam.Name == varName {
-			if varToParam.ParameterName != "" {
-				return varToParam.ParameterName, nil
-			}
-			break
+// convertVarToParam converts a project variable to its equivalent parameter
+// name and value. In particular, it validates that the variable name and value
+// fits within parameter constraints and if the name or value doesn't fit in the
+// constraints, it attempts to fix minor issues where possible. The return value
+// is a valid parameter name and parameter value.
+func convertVarToParam(projectID string, pm ParameterMappings, varName, varValue string) (paramName string, paramValue string, err error) {
+	if err := validateVarNameCharset(varName); err != nil {
+		return "", "", errors.Wrapf(err, "validating project variable name '%s'", varName)
+	}
+
+	varsToParams := pm.NameMap()
+	m, ok := varsToParams[varName]
+	if ok {
+		paramName = m.ParameterName
+	} else {
+		paramName, err = createParamBasenameForVar(varName)
+		if err != nil {
+			return "", "", errors.Wrapf(err, "creating new parameter name for project variable '%s'", varName)
 		}
 	}
 
-	paramName := varName
-	if !validParamName.MatchString(paramName) {
-		return "", errors.Errorf("project variable '%s' contains invalid characters - can only contain alphanumerics, underscores, periods, and dashes", varName)
+	paramName, paramValue, err = getCompressedParamForVar(paramName, varValue)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "getting compressed parameter name and value for project variable '%s'", varName)
 	}
 
-	if strings.HasPrefix(varName, "aws") || strings.HasPrefix(paramName, "ssm") {
-		// Parameters cannot start with "aws" or "ssm", adding a prefix
-		// (arbitrarily chosen as an underscore) fixes the issue.
-		paramName = fmt.Sprintf("_%s", paramName)
+	prefix := fmt.Sprintf("%s/", projectID)
+	if !strings.Contains(paramName, prefix) {
+		paramName = fmt.Sprintf("%s%s", prefix, paramName)
+	}
+
+	if err := validateParamNameUnique(pm, varName, paramName); err != nil {
+		return "", "", errors.Wrapf(err, "validating parameter name for project variable '%s'", varName)
+	}
+
+	return paramName, paramValue, nil
+}
+
+// validParamBasename is a regexp representing the valid characters for a
+// parameter's base name (i.e. excluding any slash-delimited paths). Valid
+// characters for a basename are alphanumerics, underscores, dashes, and
+// periods.
+var validParamBasename = regexp.MustCompile(`^[a-zA-Z0-9_.-]+$`)
+
+// validateVarNameCharset verifies that a project variable name is not empty and
+// contains only valid characters. It returns an error if it it's empty or
```

Comment:
nit: `It returns an error if it's empty`

Reply:
Fixed.